File: kotlin/bc-envelope/src/main/kotlin/com/blockchaincommons/bcenvelope/EnvelopeTree.kt
package com.blockchaincommons.bcenvelope

import com.blockchaincommons.bccomponents.Digest
import com.blockchaincommons.bccomponents.DigestProvider
import com.blockchaincommons.bcur.UREncodable

/**
 * Format for displaying digests in the tree representation.
 */
enum class DigestDisplayFormat {
    /** Display a shortened version of the digest (first 8 characters). */
    Short,
    /** Display the full digest for each element in the tree. */
    Full,
    /** Display a ur:digest UR for each element in the tree. */
    UR,
}

/**
 * Options for tree-formatted output of an envelope.
 */
class TreeFormatOpts(
    val hideNodes: Boolean = false,
    val highlightingTarget: Set<Digest> = emptySet(),
    val context: FormatContextOpt = FormatContextOpt.Global,
    val digestDisplay: DigestDisplayFormat = DigestDisplayFormat.Short,
)

// -- Tree format extensions --

/** Returns a tree-formatted string representation of the envelope. */
fun Envelope.treeFormat(): String = treeFormatOpt(TreeFormatOpts())

/** Returns a tree-formatted string representation with options. */
fun Envelope.treeFormatOpt(opts: TreeFormatOpts): String {
    val elements = mutableListOf<TreeElement>()
    val visitor = { envelope: Envelope, level: Int, incomingEdge: EdgeType, _: Unit ->
        val elem = TreeElement(
            level = level,
            envelope = envelope,
            incomingEdge = incomingEdge,
            showId = !opts.hideNodes,
            isHighlighted = opts.highlightingTarget.contains(envelope.digest()),
        )
        elements.add(elem)
        Pair(Unit, false)
    }
    walk(opts.hideNodes, Unit, visitor)

    val formatElements = { elems: List<TreeElement>, ctx: FormatContext ->
        elems.joinToString("\n") { it.toFormattedString(ctx, opts.digestDisplay) }
    }

    return when (opts.context) {
        is FormatContextOpt.None -> {
            val defaultCtx = FormatContext()
            formatElements(elements, defaultCtx)
        }
        is FormatContextOpt.Global -> withFormatContext { ctx ->
            formatElements(elements, ctx)
        }
        is FormatContextOpt.Custom -> formatElements(elements, opts.context.context)
    }
}

/** Returns a text representation of the envelope's digest. */
fun Envelope.shortId(format: DigestDisplayFormat = DigestDisplayFormat.Short): String =
    when (format) {
        DigestDisplayFormat.Short -> digest().shortDescription()
        DigestDisplayFormat.Full -> digest().hex
        DigestDisplayFormat.UR -> (digest() as UREncodable).urString()
    }

/**
 * An element in the tree representation of an envelope.
 */
private class TreeElement(
    val level: Int,
    val envelope: Envelope,
    val incomingEdge: EdgeType,
    val showId: Boolean,
    val isHighlighted: Boolean,
) {
    fun toFormattedString(context: FormatContext, digestDisplay: DigestDisplayFormat): String {
        val parts = mutableListOf<String>()
        if (isHighlighted) parts.add("*")
        if (showId) parts.add(envelope.shortId(digestDisplay))
        incomingEdge.label()?.let { parts.add(it) }
        parts.add(envelope.summary(40, context))
        val line = parts.joinToString(" ")
        val indent = " ".repeat(level * 4)
        return "$indent$line"
    }
}

File: typescript/bc-envelope/src/envelope-tree.ts
import { type Digest } from '@bc/components';

import type { Envelope } from './envelope.js';
import { EdgeType, edgeTypeLabel } from './edge-type.js';
import type { FormatContext, FormatContextOpt } from './format-context.js';
import { FormatContextOpts, withFormatContext } from './format-context.js';

export enum DigestDisplayFormat {
    Short = 'short',
    Full = 'full',
    UR = 'ur',
}

export interface TreeFormatOpts {
    hideNodes?: boolean;
    highlightingTarget?: Set<Digest>;
    context?: FormatContextOpt;
    digestDisplay?: DigestDisplayFormat;
}

interface TreeElement {
    level: number;
    envelope: Envelope;
    incomingEdge: EdgeType;
    showId: boolean;
    isHighlighted: boolean;
}

function digestIn(target: Set<Digest>, digest: Digest): boolean {
    for (const item of target) {
        if (item.equals(digest)) {
            return true;
        }
    }
    return false;
}

function shortId(envelope: Envelope, format: DigestDisplayFormat): string {
    const digest = envelope.digest();
    if (format === DigestDisplayFormat.Short) {
        return digest.shortDescription;
    }
    if (format === DigestDisplayFormat.Full) {
        return digest.hex();
    }
    return digest.urString();
}

function toFormattedString(element: TreeElement, context: FormatContext, digestDisplay: DigestDisplayFormat): string {
    const parts: string[] = [];
    if (element.isHighlighted) {
        parts.push('*');
    }
    if (element.showId) {
        parts.push(shortId(element.envelope, digestDisplay));
    }
    const label = edgeTypeLabel(element.incomingEdge);
    if (label != null) {
        parts.push(label);
    }
    parts.push(element.envelope.summary(40, context));
    return `${' '.repeat(element.level * 4)}${parts.join(' ')}`;
}

export function treeFormat(envelope: Envelope): string {
    return treeFormatOpt(envelope, {});
}

export function treeFormatOpt(envelope: Envelope, opts: TreeFormatOpts): string {
    const hideNodes = opts.hideNodes ?? false;
    const highlightingTarget = opts.highlightingTarget ?? new Set<Digest>();
    const context = opts.context ?? FormatContextOpts.global();
    const digestDisplay = opts.digestDisplay ?? DigestDisplayFormat.Short;

    const elements: TreeElement[] = [];
    envelope.walk(hideNodes, undefined, (item, level, incomingEdge, state) => {
        elements.push({
            level,
            envelope: item,
            incomingEdge,
            showId: !hideNodes,
            isHighlighted: digestIn(highlightingTarget, item.digest()),
        });
        return [state, false];
    });

    const formatElements = (formatContext: FormatContext): string => {
        return elements.map((item) => toFormattedString(item, formatContext, digestDisplay)).join('\n');
    };

    if (context.type === 'none') {
        return formatElements(withFormatContext((global) => global));
    }
    if (context.type === 'global') {
        return withFormatContext((global) => formatElements(global));
    }
    return formatElements(context.context);
}
